keps/sig-api-machinery/4222-cbor-serializer/README.md

```diff
@@ -0,0 +1,1274 @@
+
+> the self-delimiting nature of the CBOR encoding means that there are no two
+> well-formed CBOR encoded data items where one is a prefix of the other
+
+In other words, CBOR (and the existing JSON framing) can stream directly to and
+from the wire without incurring additional copies on both sides of the
+connection. If an encoding fails or is otherwise not completely received on the
+other end, the fragment that _is_ received will not be well-formed and will
+produce a decode error.
+
+### Negotiation
+
+Proactive content negotiation will be supported for clients that want to receive
+CBOR-encoded responses using the MIME type “application/cbor” in the Accept
+request header. For compatibility with API servers that don’t support CBOR,
+clients should also accept “application/json” (with a lower quality factor) and
+choose the appropriate decoder based on the Content-Type response header.
+
+Similarly, streaming requests should use the MIME type for CBOR Sequences,
+“application/cbor-seq”.
+
+CBOR will not be a supported encoding for JSON Patches, JSON Merge Patches,
+Strategic Merge Patches, or Server-Side Apply configurations.
+
+Clients can send CBOR-encoded request bodies with the appropriate Content-Type
+to API servers that support CBOR. API servers that don’t support CBOR will
+return status 415 (Unsupported Media Type). In client-go, for alpha, when a
+RESTClient configured to encode requests with CBOR receives a 415, it will
+permanently (for the life of the RESTClient) fall back to JSON for subsequent
+requests. For GA, this fallback behavior will be changed to operate on a
+per-(method, target resource) basis, and to consider acceptable fallback
+content-types based on the value of the Accept header in a 415 response, [as
+described in RFC 9110](https://httpwg.org/specs/rfc9110.html#status.415).
+
+### Client Enablement
+
+Clients can be explicitly configured to prefer CBOR as a request encoding as
+they can today be configured to prefer Protobuf or JSON. In client-go, this
+involves setting the `ContentType` field of `rest.ClientContentConfig`. The
+default request content-type will remain JSON for a period of time post-GA; a
+minimum of two minor versions, so that the oldest kube-apiserver within the
+supported kubectl version skew will have CBOR support. The supported version
+skew for aggregated API servers is much wider (infinite?). Encoding and decoding
+resources from aggregated API servers that don't support CBOR will rely on the
+content-type negotation mechanisms described above.
+
+CBOR enablement will be feature-gated in API servers. Client-go will provide a
+runtime gating mechanism for CBOR request encoding as follows:
+
+- An environment variable will be read at package initialization time with
+  recognized values:
+  - "default": The default request content-type (if not explicitly configured)
+    becomes application/cbor rather than application/json.
```

Comment:
I'm guessing more like `application/cbor,application/json`, the language here doesn't make it super clear.